File: app/src/main/java/com/rowland/delivery/features/dash/data/model/CategoryRepository.java
package com.rowland.delivery.features.dash.data.model;

import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.SetOptions;
import com.rowland.delivery.features.dash.domain.models.Category;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import durdinapps.rxfirebase2.RxFirestore;
import io.reactivex.Flowable;
import io.reactivex.Single;

/**
 * Created by Rowland on 5/5/2018.
 */

public class CategoryRepository {

    private FirebaseFirestore mFirebaseFirestone;

    @Inject
    public CategoryRepository(FirebaseFirestore firebaseFirestone) {
        this.mFirebaseFirestone = firebaseFirestone;
    }

    public Flowable<List<Category>> getCategories(String userUID) {
        CollectionReference categoryCollectionRef = mFirebaseFirestone.collection("categories");
        Query query = categoryCollectionRef.whereEqualTo(String.format("members.%s", userUID), true);
        return RxFirestore.observeQueryRef(query, Category.class);
    }


    public Single<Category> createCategory(Category category, String userUID) {
        DocumentReference documentReference = mFirebaseFirestone.collection("categories").document(category.getName());
        documentReference.set(category, SetOptions.merge()).addOnSuccessListener(aVoid -> {
            Map<String, Object> members = new HashMap<>();
            members.put(String.format("members.%s", userUID), true);
            documentReference.update(members);
        });

        return RxFirestore.getDocument(documentReference, Category.class).toSingle();
    }
}

File: app/src/main/java/com/rowland/delivery/features/dash/data/model/category/CategoryRepository.kt
package com.rowland.delivery.features.dash.data.model.category

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.firestore.SetOptions
import com.rowland.delivery.features.dash.domain.contracts.ICategoryRepository
import com.rowland.delivery.features.dash.domain.models.category.CategoryEntity
import com.rowland.delivery.features.dash.presentation.tools.snapshots.DocumentWithIdSnapshotMapper
import durdinapps.rxfirebase2.RxFirestore
import io.reactivex.Flowable
import io.reactivex.Single
import java.util.*
import javax.inject.Inject

/**
 * Created by Rowland on 5/5/2018.
 */

class CategoryRepository @Inject
constructor(private val mFirebaseFirestone: FirebaseFirestore) : ICategoryRepository {

    override fun getCategories(userUid: String): Flowable<List<CategoryEntity>> {
        val categoryCollectionRef = mFirebaseFirestone.collection("categories")
        val query = categoryCollectionRef.whereEqualTo(String.format("merchants.%s", userUid), true)
        return RxFirestore.observeQueryRef(query, DocumentWithIdSnapshotMapper.listOf(CategoryEntity::class.java) as io.reactivex.functions.Function<QuerySnapshot, List<CategoryEntity>>)
    }

    override fun createCategory(categoryEntity: CategoryEntity, userUid: String): Single<CategoryEntity> {
        val documentReference = mFirebaseFirestone.collection("categories").document(categoryEntity.name!!)
        documentReference.set(categoryEntity, SetOptions.merge()).addOnSuccessListener { _ ->
            val members = HashMap<String, Any>()
            members[String.format("merchants.%s", userUid)] = true
            documentReference.update(members)
        }
        return RxFirestore.getDocument(documentReference, CategoryEntity::class.java).toSingle()
    }
}
